### DISCUSSION TITLE
vivado_hls missing when using hls_model.build()

### DISCUSSION THREAD

**Discussion by DrBwts** [fastmachinelearning/hls4ml#1087] [Q&A] [answered]
Hi,
Ubuntu 24.04
Vivado 2024.1
Installed the HLS4ML environment as detailed on the tutorials page.

I'm trying to go through the tutorials but run into a problem in the first notebook. Everything runs fine until the second to last cell,

`hls_model.build(csim=False)`

Which throws the following error,

```
Exception                                 Traceback (most recent call last)
Cell In[15], line 1
----> 1 hls_model.build(csim=False)
File ~/anaconda3/envs/hls4ml-tutorial/lib/python3.10/site-packages/hls4ml/model/graph.py:854, in ModelGraph.build(self, **kwargs)
    850 if not os.path.exists(self.config.get_output_dir()):
    851     # Assume the project wasn't written before
    852     self.write()
--> 854 return self.config.backend.build(self, **kwargs)

File ~/anaconda3/envs/hls4ml-tutorial/lib/python3.10/site-packages/hls4ml/backends/vivado/vivado_backend.py:203, in VivadoBackend.build(self, model, reset, csim, synth, cosim, validation, export, vsynth, fifo_opt)
    201     found = os.system('command -v vivado_hls > /dev/null')
    202     if found != 0:
--> 203         raise Exception('Vivado HLS installation not found. Make sure "vivado_hls" is on PATH.')
    205 curr_dir = os.getcwd()
    206 os.chdir(model.config.get_output_dir())

Exception: Vivado HLS installation not found. Make sure "vivado_hls" is on PATH.
```

I have searched the Xilinx install folders but just cant find `vivado_hls` anywhere. My current suspicion is that this component has been integrated somewhere within Vitis & now has a new name.

Anyway around this? Is is possible to get older versions of Vivado installed? If so which version are you using?

Any plans to update to the latest version?

**Comment by jmitrevs** [comment]
Vivado_HLS was replaced by Vitis_HLS in newer releases. You need to use the "Vitis" backend.

**Comment by DrBwts** [comment]
Is there a way of doing that from the notebook?

**Reply by jmitrevs** [reply]
Lots of hls4ml functions have a backend argument that you can fill. If you are looking at the tutorial, see some of the other branches, like oneAPI, which uses the oneAPI backend. (There may also be a Vitis branch, I forget.)

**Comment by JanFSchulte** [comment]
There is a Vitis branch of the tutorial: https://github.com/fastmachinelearning/hls4ml-tutorial/tree/vitis. The only real difference is the choice of the backend in the `config_from_keras_model` and `convert_from_keras_model` calls, as well as making sure the Vitis HLS installation is on the PATH. In this branch, it is cardcoded as 

`os.environ['PATH'] = '/opt/Xilinx//Vitis_HLS/2024.1/bin:' + os.environ['PATH']`

to work with a specific docker container, so you would need to adjust that for your installation.

**Reply by DrBwts** [reply]
Thanks will give it a go.

**Comment by DrBwts** [comment]
Unfortunately that gives me the exact same error.

**Reply by JanFSchulte** [reply]
Not much we can do to help I'm afraid. We did run a tutorial based on that branch last Friday and it works fine as long as Vitis HLS is installed and path is set correctly. So you will need to figure out your setup and make sure that is the case.

**Reply by DrBwts** [reply]
The odd thing is that the error is still about vivado_hls,

```
File ~/anaconda3/envs/hls4ml-tutorial/lib/python3.10/site-packages/hls4ml/backends/vivado/vivado_backend.py:203, in VivadoBackend.build(self, model, reset, csim, synth, cosim, validation, export, vsynth, fifo_opt)
    201     found = os.system('command -v vivado_hls > /dev/null')
    202     if found != 0:
--> 203         raise Exception('Vivado HLS installation not found. Make sure "vivado_hls" is on PATH.')
    205 curr_dir = os.getcwd()
    206 os.chdir(model.config.get_output_dir())

Exception: Vivado HLS installation not found. Make sure "vivado_hls" is on PATH.
```

**Reply by JanFSchulte** [reply]
Yeah, that looks like the Vivado backend is still being used. Can you double check the notebook you are running and see if the Vitis backendend is used everywhere?

**Reply by DrBwts** [reply]
```
config = hls4ml.utils.config_from_keras_model(model, granularity='model')
hls_model = hls4ml.converters.convert_from_keras_model(model, hls_config=config, output_dir='model_1/hls4ml_prj', part='xcu250-figd2104-2L-e')
```
 Is the only code I could find that deals with hls4ml model configuration.

**Reply by DrBwts** [reply]
I think its an issue with `git clone` it wont let me clone `.../tree/vitis` 
```

git clone https://github.com/fastmachinelearning/hls4ml-tutorial/tree/vitis.git
Cloning into 'vitis'...
fatal: repository 'https://github.com/fastmachinelearning/hls4ml-tutorial/tree/vitis.git/' not found
```


`Code` gives https://github.com/fastmachinelearning/hls4ml-tutorial.git which is the main branch

**Reply by jmitrevs** [reply, marked as answer]
You need to do `git checkout vitis` to switch the branches.

**Reply by DrBwts** [reply]
Yes that's done it (my git fu is weak)!

Thank you so much for your time.

**Reply by kalahel** [reply]
Hi I have the same problem and I'm on the vitis branch.

How do you set up correctly the path to vitis hls ?

**Reply by JanFSchulte** [reply]
Hi! 

You have to add the path the `bin` directory of your Vitis installation to the `PATH`  environment variable. 

That can be done either by running `source settings64.sh` in the `bin` folder, after which the path is available as it's own environment variable and can be added to the `PATH` with

```
os.environ['PATH'] = os.environ['XILINX_VITIS'] + '/bin:' + os.environ['PATH']
```

Or alternatively you can directly add the explicit path to wherever you have installed Vitis, for example in case of the tutorial:

```
os.environ['PATH'] = '/opt/Xilinx//Vitis_HLS/2024.1/bin:' + os.environ['PATH']
```
